lib/src/main/java/com/knuddels/jtokkit/GptBytePairEncoding.java

```diff
@@ -5,321 +5,309 @@
-				// the merged byte pair. We use the skip parameter of the getRank function to calculate the rank of, in our
-				// example, "t" + "o" + "r" and "c" + "t" + "o". The assumption made in the OpenAI implementation is that
-				// removing first thrashes the cache, so it's better to calculate the rank of the byte pairs that are
-				// affected by the merge before removing the merged byte pair. I did not verify, if this is actually the
-				// case in java.
-				parts.get(minRankIndex).rank = getRank(piece, parts, minRankIndex, 1).orElse(Integer.MAX_VALUE);
-				if (minRankIndex > 0) {
-					parts.get(minRankIndex - 1).rank = getRank(piece, parts, minRankIndex - 1, 1).orElse(Integer.MAX_VALUE);
-				}
-
-				parts.remove(minRankIndex + 1);
-			} else {
-				break;
-			}
-		}
-
-		/*
-		 * piece:  vec tor
-		 * index:  0   3   6
-		 * ranks:  inf inf inf
-		 */
-		final List<Integer> out = new ArrayList<>();
-		for (int i = 0; i < parts.size() - 1; i++) {
-			out.add(encoder.encode(piece.getBytesBetween(parts.get(i).index, parts.get(i + 1).index)));
-		}
-		return out;
-	}
-
-	private boolean maxTokenCountReached(final Integer maxTokenCount, final int tokenCount) {
-		return maxTokenCount != null && maxTokenCount.compareTo(tokenCount) <= 0;
-	}
-
-	private boolean maxTokenCountNotReached(final Integer maxTokenCount, final int tokenCount) {
-		return !maxTokenCountReached(maxTokenCount, tokenCount);
-	}
-
-	private Optional<Integer> getRank(
-			final ImmutableByteArray piece,
-			final List<PieceIndexToRank> parts,
-			final int startIndex,
-			final int skip
-	) {
-		if (startIndex + skip + 2 >= parts.size()) {
-			return Optional.empty();
-		}
-
-		final int pieceStartIndex = parts.get(startIndex).index;
-		final int pieceEndIndex = parts.get(startIndex + skip + 2).index;
-		final ImmutableByteArray encoderIndex = piece.getBytesBetween(pieceStartIndex, pieceEndIndex);
-
-		return encoder.encodeIfPresent(encoderIndex);
-	}
-
-	private byte[] decodeToken(final int token) {
-		final Optional<ImmutableByteArray> decodedToken = encoder.decodeIfPresent(token);
-		if (decodedToken.isPresent()) {
-			return decodedToken.get().getRawArray();
-		}
-
-		final Optional<String> decodedSpecialToken = specialTokensEncoder.decodeIfPresent(token);
-		if (decodedSpecialToken.isPresent()) {
-			return decodedSpecialToken.get().getBytes(StandardCharsets.UTF_8);
-		}
-
-		throw new IllegalArgumentException("Unknown token for decoding: " + token);
-	}
-
-	private static class PieceIndexToRank {
-		private final int index;
-		private int rank;
-
-		public PieceIndexToRank(final int index, final int rank) {
-			this.index = index;
-			this.rank = rank;
-		}
-	}
+class GptBytePairEncoding implements Encoding {
+    private final String name;
+    private final Pattern pattern;
+    private final TokenEncoder encoder;
+    private final SpecialEncoder specialEncoder;
+    private final Map<Integer, byte[]> encodedToDecoded;
```

Comment:
TokenEncoder is only used for encoding now - so we can eliminate the types

Reply:
In my original over-optimized version there were multiple TokenEncoders - when the byte count was < Long.BYTES, we stored them in a primitive long and used a primitive map with a long key instead - that enabled squeezing the last few drops, since short tokens are the most common ones and they're a lot faster since no byte arrays are present anymore. But I haven't committed that yet, so probably I should merge it back with TokenEncoder for now and maybe split it out again if there's a PR #77

Reply:
Done